main/server.go

```diff
@@ -0,0 +1,305 @@
+//
+// Every method below gives more information about every API call, its
+// parameters, and its results.
+package main
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strconv"
+
+	"github.com/GoogleCloudPlatform/gcloud-golang-todos/todo"
+	"github.com/gorilla/mux"
+	"golang.org/x/net/context"
+	"google.golang.org/appengine"
+	"google.golang.org/appengine/log"
+)
+
+const PathPrefix = "/api/todos"
+const SlashedPathPrefix = PathPrefix + "/"
+
+var learnJson []byte
+
+func init() {
+	r := mux.NewRouter()
+	r.HandleFunc(PathPrefix,
+		errorHandler(ListTodos)).Methods("GET")
+	r.HandleFunc(PathPrefix,
+		errorHandler(NewTodo)).Methods("POST")
+	r.HandleFunc(PathPrefix,
+		errorHandler(DeleteCompletedTodos)).Methods("DELETE")
+	r.HandleFunc(SlashedPathPrefix+"{id}",
+		errorHandler(GetTodo)).Methods("GET")
+	r.HandleFunc(SlashedPathPrefix+"{id}",
+		errorHandler(UpdateTodo)).Methods("PUT")
+	r.HandleFunc(SlashedPathPrefix+"{id}",
+		errorHandler(DeleteTodo)).Methods("DELETE")
+	http.Handle("/", r)
+	http.HandleFunc("/learn.json", RenderSidebarJson)
+	http.HandleFunc("/api", IsApiEnabled)
+}
+
+// badRequest is handled by setting the status code in the reply to StatusBadRequest.
+type badRequest struct{ error }
+
+// notFound is handled by setting the status code in the reply to StatusNotFound.
+type notFound struct{ error }
+
+// IsApiEnabled writes an HTTP 200 indicating that the TodoMvc API is enabled for this app.
+func IsApiEnabled(w http.ResponseWriter, r *http.Request) {
+	w.WriteHeader(http.StatusOK)
+}
+
+// errorHandler wraps a function returning an error by handling the error and returning a http.Handler.
+// If the error is of the one of the types defined above, it is handled as described for every type.
+// If the error is of another type, it is considered as an internal error and its message is logged.
+func errorHandler(f func(w http.ResponseWriter, r *http.Request, c context.Context) error) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		c := appengine.NewContext(r)
+		err := f(w, r, c)
+		if err == nil {
+			return
+		}
+		switch err.(type) {
+		case badRequest:
+			http.Error(w, err.Error(), http.StatusBadRequest)
+		case notFound:
+			http.Error(w, "todo not found", http.StatusNotFound)
+		default:
+			log.Errorf(c, "internal exception %v", err)
+			http.Error(w, "oops", http.StatusInternalServerError)
+		}
+	}
+}
+
+// ListTodos handles GET requests on /todos.
+// It requires no parameters and returns a list of todos.
+//
+// Example:
+//
+//   req: GET /todos/
+//   res: 200 [
+//          {"id": 1, "title": "Learn Go", "completed": false},
+//          {"id": 2, "title": "Buy bread", "completed": true}
+//        ]
+func ListTodos(w http.ResponseWriter, r *http.Request, c context.Context) error {
+	res, err := todo.All(c)
+	if err != nil {
+		log.Errorf(c, "ListTodos: %v", err)
+		return err
+	}
+	w.Header().Set("Content-Type", "application/json; charset=utf-8")
```

Comment:
Would you prefer that over modifying errorHandler (perhaps with a better name) to be 

func errorHandler(f func(r *http.Request) ({}interface, error)) http.HandlerFunc {
    return func(w http.ResponseWriter, r *http.Request) {
        v, err := f(r, etc, etc)
        // handle err
        w.Header().Set(json stuff)
        err = json.NewEncoder(w).Encode(v)
        // handle err
    }
}

Basically, I haven't been able to decide whether it's more desirable to bubble interface{} stuff "down" the call stack (as in your example), or "up", as in mine. Love to hear what you think.


I've also seen some third-party code that uses httptest.ResponseRecorder to do this, but it seems...dirty.